### PR TITLE
feat(guided remediation): add `--upgrade-config` flag

### LINKED ISSUE #1177: Allow control over allowed individual package upgrade versions in non-interactive guided remediation
Currently, `osv-scanner fix` has two flags:
- `--disallow-major-upgrades` (bool), which prevents major updates in all packages
- `--disallow-package-upgrades` (list of string), which prevents all updates to the listed packages.

We should replace these flags with per-package allowed update levels e.g. `--allowed-upgrades=foo=none,bar=major,baz=minor`.

Syntax might be a bit tricky and needs some thought:
- Should this be `allowed-upgrades` or `disallowed-upgrades`?
- Is `pkg1=type1,pkg2=type2` okay syntax for all ecosystems? (i.e. does any ecosystem allow for '=' in a package name?)
- How to apply rules to all packages (e.g. `all=major` could refer to the package named `all`)
- What can be allowed.? e.g. `major`/`minor`/`patch`/`none`
  - Consider what this would mean for non-semver ecosystems
  - Maybe allow for some regex rules to deal with e.g. [com.google.guava:guava](https://deps.dev/maven/com.google.guava%3Aguava) having versions that end with `-jre` and `-android`

### PR DESCRIPTION
**Pull request by michaelkedar**
closes #1177

- Adds `--upgrade-config` flag for configuring allowed upgrades on a per-package basis.
- Hide & deprecate previous `--disallow-major-upgrades` and `--disallow-package-upgrades` flags.
- Update docs & example script to use new flag.

### CODE REVIEW

File: docs/guided-remediation.md [hunk 1]
```diff
@@ -224,10 +224,24 @@ A vulnerability is only considered if it satisfies all the conditions set by the
...
+- `--upgrade-config=<[package-name:]level>` Sets the maximum upgrade level allowed for a package. Can be repeated for multiple packages.
+
+  `level` is the SemVer component to allow updates to, can be one of `major`, `minor`, `patch`, or `none`. e.g. If a package was at version `1.2.3`
+
+  - `major` allows for updates to any version `>=1.2.3`
+  - `minor` allows for updates `>=1.2.3, <2.0.0`
+  - `patch` allows for updates `>=1.2.3, <1.3.0`
+  - `none` disallows any updates
+
+  If `package-name:` is omitted, `level` is applied to all packages. Default is `--upgrade-config=major`.
```
Comment (oliverchang) [inline docs/guided-remediation.md:238]: Can we add some detail to describe precedence? 

i.e. our behavior is to take the most specific setting right? if a specific package is specified in a config, that takes precedence over a global config.

**Review by oliverchang — APPROVED** [review]
nice!

File: docs/guided-remediation.md (hunk 1 above)
Comment (cuixq) [reply docs/guided-remediation.md:238]: +1 - it is not clear to me whether the specific settings overwrite the global ones until I see the example usage.

File: internal/remediation/upgrade/config_test.go
Comment (cuixq) [inline internal/remediation/upgrade/config_test.go:1]: maybe add some test for Maven names like `foo:bar` considering the presence of `:` in the name

**Review by cuixq — COMMENTED** [review]
LGTM!

File: docs/guided-remediation.md (hunk 1 above)
Comment (michaelkedar) [reply docs/guided-remediation.md]: Added some explanation w.r.t precedence here

File: internal/remediation/upgrade/config_test.go
Comment (michaelkedar) [reply internal/remediation/upgrade/config_test.go:1]: The logic for parsing the `pkg:level` strings is in `parseUpgradeConfig()` in `fix/main.go` - I need to work out a test for that function

Comment (michaelkedar) [reply internal/remediation/upgrade/config_test.go:1]: This as a test case in `fix/main_test.go`

**Review by cuixq — APPROVED** [review]

**Pull request merged**